### ISSUE TITLE
Error in Sys.setenv(PATH = new_path) : wrong length for argument 

### ISSUE THREAD

**Issue by philiporlando** [rstudio/reticulate#1189]
I am trying to interface R with a [proprietary conda environment](https://pro.arcgis.com/en/pro-app/latest/arcpy/get-started/what-is-arcpy-.htm) maintained by [Esri](https://en.wikipedia.org/wiki/Esri). This environment includes the ArcPy site package which contains a bunch of geospatial operations that I am unfortunately dependent on for work.

I am not able to import any python modules via reticulate.

```r
import("pandas")
# Error in Sys.setenv(PATH = new_path) : wrong length for argument
# x
# 1. +-reticulate::import("pandas")
# 2. | \-reticulate::ensure_python_initialized(required_module = module) at reticulate/R/import.R:111:4
# 3. |   \-reticulate::initialize_python(required_module, use_environment) at reticulate/R/package.R:56:2
# 4. |     +-base::local(...) at reticulate/R/package.R:113:2
# 5. |     | \-base::eval.parent(substitute(eval(quote(expr), envir)))
# 6. |     |   \-base::eval(expr, p)
# 7. |     |     \-base::eval(expr, p)
# 8. |     \-base::eval(...)
# 9. |       \-base::eval(...)
# 10. |         \-reticulate::py_discover_config(required_module, use_environment) at reticulate/R/package.R:116:4
# 11. |           \-reticulate::python_config(python_version, required_module, forced = "RETICULATE_PYTHON") at reticulate/R/config.R:186:4
# 12. |             \-reticulate::python_munge_path(python) at reticulate/R/config.R:643:2
# 13. |               \-base::Sys.setenv(PATH = new_path) at reticulate/R/config.R:573:4
# 14. \-global `<fn>`()
```

I am using the latest development version of `reticulate` (`reticulate_1.24-9000`) and have set my path to my `arcgispro-py3-clone` conda environment via the `RETICULATE_PYTHON` environment variable within my `.Renviron` file.

```r
Sys.getenv("RETICULATE_PYTHON")
# [1] "C:/Program Files/ArcGIS/Pro/bin/Python/envs/arcgispro-py3-clone/python.exe"
```

Note, the space in this file path between `Program` and `Files`. This has caused issues with reticulate in the past (#1149). That being said, I think something different is happening with the latest version. Please review my debugging efforts below and let me know if I can provide any additional information to help resolve this issue. Thank you.

1. `import()` breaks here `import.R:111` when calling `ensure_python_initialized()`.
2. `ensure_python_initialized()` breaks here `package.R:56` when calling `initialize_python()`.
3. `initialize_python()` breaks here `package.R:116` when calling `py_discover_config()`.
4. `py_discover_config()` breaks here `config.R:186` when calling `python_config()`.
5.  `python_config()` breaks here `config.R:643` when calling `python_munge_path()`.
6.  `python_munge_path()` breaks here `config.R:573` when calling `Sys.setenv("PATH" = new_path)`

The `new_path` object is a list containing the following elements:
```r
# [1] ""                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                      
# [2] "CommandNotFoundError: 'activate'"                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                      
# [3] ""                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                      
# [4] "C:\\Program Files\\ArcGIS\\Pro\\bin\\Python\\envs\\arcgispro-py3-clone\\Library\\bin;C:\\Program Files\\ArcGIS\\Pro\\bin;C:\\Program Files\\R\\R-4.1.3\\bin\\x64;C:\\Program Files\\Common Files\\Oracle\\Java\\javapath_target_483022281;C:\\Program Files (x86)\\Common Files\\Oracle\\Java\\javapath_target_1291393390;C:\\orant\\Ora12\\bin;C:\\Windows\\System32;C:\\Windows;C:\\Windows\\System32\\wbem;C:\\Windows\\System32\\WindowsPowerShell\\v1.0;C:\\Windows\\System32\\OpenSSH;C:\\Program Files (x86)\\CheckPoint\\Endpoint Security\\Endpoint Common\\bin;C:\\Program Files (x86)\\Microsoft SQL Server\\Client SDK\\ODBC\\130\\Tools\\Binn;C:\\Program Files (x86)\\Microsoft SQL Server\\140\\Tools\\Binn;C:\\Program Files (x86)\\Microsoft SQL Server\\140\\DTS\\Binn;C:\\Program Files (x86)\\Microsoft SQL Server\\140\\Tools\\Binn\\ManagementStudio;C:\\Program Files\\SASHome\\Secure\\ccme4;C:\\Program Files\\MiKTeX\\miktex\\bin\\x64;C:\\Program Files\\ArcGIS\\Pro\\bin\\Python\\envs\\arcgispro-py3-clone;C:\\Program Files\\ArcGIS\\Pro\\bin\\Python\\envs\\arcgispro-py3-clone\\Scripts;;C:\\Program Files\\WindowsPowerShell\\Scripts\\HP.ClientScriptLibrary;C:\\Users\\A042474\\AppData\\Local\\Microsoft\\WindowsApps;C:\\Program Files\\R\\R-4.1.3\\bin;C:\\Program Files\\R\\R-4.1.3\\bin\\x64;C:\\Users\\A042474\\AppData\\Local\\Programs\\Microsoft VS Code\\bin;C:\\local\\pandoc-2.13;C:\\Users\\A042474\\AppData\\Roaming\\TinyTeX\\bin\\win32;C:\\Users\\A042474\\AppData\\Local\\Atlassian\\SourceTree\\git_local;C:\\Program Files\\ArcGIS\\Pro\\bin\\Python\\envs\\arcgispro-py3-clone;C:\\Program Files\\ArcGIS\\Pro\\bin\\Python\\envs\\arcgispro-py3-clone\\Scripts"
```

The `wrong length for argument` error makes sense because we are trying to set a new environment variable called `PATH` and erroneously assigning it a list of paths instead of a single value. 

The `new_path` object is returned by calling `conda_run2` within `python_munge_path`:
```r
new_path <- conda_run2(
  "python",
  c("-c", shQuote("import os; print(os.environ['PATH'])")),
  conda = conda_info$conda,
  envname = conda_info$root,
  intern = TRUE
)

# this is what the conda_info values are
conda_info$root
# $conda
# [1] "C:/Program Files/ArcGIS/Pro/bin/Python/Scripts/conda.exe"
# 
# $root
# [1] "C:/Program Files/ArcGIS/Pro/bin/Python/envs/arcgispro-py3-clone"
```
7. Taking a closer look at `conda_run2()`, the `is_windows()` method is correctly returning `TRUE` when called from my machine, which results in the `conda_run2_windows()` being called.
8. The `conda_run2_windows()` method is being passed the following parameters:
```r
cmd <- "python"
args <- c("-c", shQuote("import os; print(os.environ['PATH'])"))
conda = "C:/Program Files/ArcGIS/Pro/bin/Python/Scripts/conda.exe"
envname = "C:/Program Files/ArcGIS/Pro/bin/Python/envs/arcgispro-py3-clone"
cmd_line = paste(shQuote(cmd), paste(args, collapse = " "))
intern = TRUE
echo = !intern
```
9. The `conda` param is normalized here `conda.R:836`:
```r
conda <- normalizePath(conda_binary(conda))
conda
# [1] "C:\\Program Files\\ArcGIS\\Pro\\bin\\Python\\Scripts\\conda.exe"
```
10. The `envname` is not identical to `"base"`, so `condaenv_resolve(envname)` is called here `conda.R:841`:
```r
envname <- condaenv_resolve(envname)
envname
# [1] "C:/Program Files/ArcGIS/Pro/bin/Python/envs/arcgispro-py3-clone"
```
11.  Forward slashes are detected within `envname` here `conda.R:843` and `normalizePath()` is called:
```r
envname <- normalizePath(envname)
envname
# [1] "C:\\Program Files\\ArcGIS\\Pro\\bin\\Python\\envs\\arcgispro-py3-clone"
```
12. A batch file is created here `conda.R:846` that contains the following commands:
```
@echo off
CALL "C:\Program Files\ArcGIS\Pro\bin\Python\Scripts\conda.exe" activate "C:\Program Files\ArcGIS\Pro\bin\Python\envs\arcgispro-py3-clone"
"python" -c "import os; print(os.environ['PATH'])"
```
13. This batch file is executed using `shell()` here `conda.R:854`. I think this step is where the error is occurring! The output of this `shell()` call returns the same list that `conda_run2()` was returning above:
```r
# [1] ""                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                      
# [2] "CommandNotFoundError: 'activate'"                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                      
# [3] ""                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                      
# [4] "C:\\Program Files\\ArcGIS\\Pro\\bin\\Python\\envs\\arcgispro-py3-clone\\Library\\bin;C:\\Program Files\\ArcGIS\\Pro\\bin;C:\\Program Files\\R\\R-4.1.3\\bin\\x64;C:\\Program Files\\Common Files\\Oracle\\Java\\javapath_target_483022281;C:\\Program Files (x86)\\Common Files\\Oracle\\Java\\javapath_target_1291393390;C:\\orant\\Ora12\\bin;C:\\Windows\\System32;C:\\Windows;C:\\Windows\\System32\\wbem;C:\\Windows\\System32\\WindowsPowerShell\\v1.0;C:\\Windows\\System32\\OpenSSH;C:\\Program Files (x86)\\CheckPoint\\Endpoint Security\\Endpoint Common\\bin;C:\\Program Files (x86)\\Microsoft SQL Server\\Client SDK\\ODBC\\130\\Tools\\Binn;C:\\Program Files (x86)\\Microsoft SQL Server\\140\\Tools\\Binn;C:\\Program Files (x86)\\Microsoft SQL Server\\140\\DTS\\Binn;C:\\Program Files (x86)\\Microsoft SQL Server\\140\\Tools\\Binn\\ManagementStudio;C:\\Program Files\\SASHome\\Secure\\ccme4;C:\\Program Files\\MiKTeX\\miktex\\bin\\x64;C:\\Program Files\\ArcGIS\\Pro\\bin\\Python\\envs\\arcgispro-py3-clone;C:\\Program Files\\ArcGIS\\Pro\\bin\\Python\\envs\\arcgispro-py3-clone\\Scripts;;C:\\Program Files\\WindowsPowerShell\\Scripts\\HP.ClientScriptLibrary;C:\\Users\\A042474\\AppData\\Local\\Microsoft\\WindowsApps;C:\\Program Files\\R\\R-4.1.3\\bin;C:\\Program Files\\R\\R-4.1.3\\bin\\x64;C:\\Users\\A042474\\AppData\\Local\\Programs\\Microsoft VS Code\\bin;C:\\local\\pandoc-2.13;C:\\Users\\A042474\\AppData\\Roaming\\TinyTeX\\bin\\win32;C:\\Users\\A042474\\AppData\\Local\\Atlassian\\SourceTree\\git_local;C:\\Program Files\\ArcGIS\\Pro\\bin\\Python\\envs\\arcgispro-py3-clone;C:\\Program Files\\ArcGIS\\Pro\\bin\\Python\\envs\\arcgispro-py3-clone\\Scripts"
```
Something that stands out is the second element of this list, the `"CommandNotFoundError: 'activate'"` error. 

Navigating to the batch file and executing it (by double clicking it) I'm getting the same error.  So it appears that there is something wrong with `conda_run2_windows()` and my `arcgispro-py3-clone` conda environment. I am getting the same error message when trying to activate this conda environment from `cmd` as well:

```
CALL "C:\Program Files\ArcGIS\Pro\bin\Python\Scripts\conda.exe" activate "C:\Program Files\ArcGIS\Pro\bin\Python\envs\arcgispro-py3-clone"
# CommandNotFoundError: 'activate'
```
![image](https://user-images.githubusercontent.com/24445174/163504935-35df34f6-cb8b-4d58-a709-456c2f8d4c43.png)

This is as far as I could take it. Please let me know if I can do anything else to help troubleshoot. It is extremely important for my job that I am able to connect R to this proprietary conda environment. That being said, I understand that it will be difficult to debug without having this version of conda on your machines.

**Comment by t-kalinowski** [comment]
I think that the issue is because the directory structure shipped by ArcGis is not the standard conda installation, and reticulate is not finding the "conda.bat" script correctly. It is using the "conda.exe" as a fallback, which can not be used in this type of context.

Can you tell me where conda.bat is in that installation? 
```
list.files("C:\\Program Files\\ArcGIS\\", recursive = TRUE, pattern = "conda\\.bat$", full.names = TRUE)
```

**Comment by philiporlando** [comment]
@t-kalinowski, thank you for your insight. I would not be surprised if the directory structure of ArcGIS is non-standard. 

````r
list.files("C:\\Program Files\\ArcGIS\\", recursive = TRUE, pattern = "conda\\.bat$", full.names = TRUE)
# [1] "C:\\Program Files\\ArcGIS\\/Pro/bin/Python/envs/arcgispro-py3-clone/Scripts/conda.bat"
# [2] "C:\\Program Files\\ArcGIS\\/Pro/bin/Python/envs/arcgispro-py3/Scripts/conda.bat" 
```

So there are two ArcGIS conda environments on my machine, `arcgispro-py3` and `arcgispro-py3-clone`. The first is the default conda environment that ships with ArcGIS Pro. The second is a cloned conda environment that I created and installed additional dependencies to. 

After jumping inside `get_python_conda_info()` I'm seeing that `dir.exists(file.path(root, "condabin"))` returns `FALSE`. As a result, `conda <- python_info_condaenv_find(root)` is called. This is where we the `C:/Program Files/ArcGIS/Pro/bin/Python/Scripts/conda.exe` is being assigned, instead of the desired paths that I linked above?

**Comment by t-kalinowski** [comment]
Do you know how ESRI intends for this environment to be activated?
Does running this commands succeed in a cmd.exe terminal?
```
CALL "C:\Program Files\ArcGIS\Pro\bin\Python\envs\arcgispro-py3\Scripts\conda.bat" activate 
```

**Comment by philiporlando** [comment]
Unfortunately, I do not fully understand how ESRI intends for this environment to be activated, or why they would structure their environment differently. 

Here is what I do know:

When ArcGIS Pro is installed, it includes a Python Command Prompt that is available from the Windows Start menu. I think this is how ESRI expects users to interact with their `arcgispro-py3` conda environment. 

![image](https://user-images.githubusercontent.com/24445174/163599536-c3eec09b-a4e4-4666-85a1-7a0f376e506b.png)

This is just a shortcut that routes to `"C:\Program Files\ArcGIS\Pro\bin\Python\Scripts\proenv.bat"` which contains the following commands:

```
@echo off
@CALL :normalizepath scripts_path "%~dp0"
@set global_path="%scripts_path%proenv.txt"
@set local_path="%localappdata%\ESRI\conda\envs\proenv.txt"

:: read the active environment name
if exist "%local_path%" (
    @set /p CONDA_NEW_ENV=<%local_path%
) else (
    @set /p CONDA_NEW_ENV=<%global_path%
)


@SET "activate_path="%scripts_path%activate.bat" "%CONDA_NEW_ENV%"
@set "deactivate_path="%scripts_path%deactivate.bat""
@set CONDA_SKIPCHECK=1
@call %activate_path%
python.exe %*
@set PY_ERRORLEVEL=%ERRORLEVEL%
@call %deactivate_path%

@exit /b %PY_ERRORLEVEL%

:normalizepath
    @set "%1=%~dpfn2"
    @exit /b
```

Within `proenv.bat` is a call to `activate.bat`, which lives in the same directory. I typically use `activate.bat` to interactively create the cloned environment. Here is what the `activate.bat` file contains:

```
@REM @ symbols in this file indicate that output should not be printed.
@REM   Setting it this way allows us to not touch the user's echo setting.
@REM   For debugging, remove the @ on the section you need to study.
@setlocal enabledelayedexpansion
@set "CONDA_NEW_ENV=%~1"

@SET "CONDA_EXE=%~dp0\..\Scripts\conda.exe"
@CALL :normalizepath scripts_path "%~dp0\..\Scripts"
@SET "CONDA_SCRIPTS=%scripts_path%"

:: this finds either --help or -h and shows the help text
@CALL ECHO "%~1"| @%SystemRoot%\System32\find.exe /I "-h" 1>NUL
@IF NOT ERRORLEVEL 1 (
    @call "%CONDA_EXE%" ..activate "cmd.exe" -h
) else (
    :: reset errorlevel to 0
    cmd /c "exit /b 0"
)

@if "%~2" == "" @goto skiptoomanyargs
    (@echo Error: did not expect more than one argument.) 1>&2
    (@echo     ^(Got %*^)) 1>&2
    @exit /b 1
:skiptoomanyargs

@if not "%~1" == "" @goto skipmissingarg
    @REM Set env to root if no arg provided
    @set CONDA_NEW_ENV=root
:skipmissingarg

@if "%CONDA_SKIPCHECK%" == "1" (
    @REM Passed in a full path, use it directly and skip conda internal validation.
    IF NOT EXIST "%CONDA_NEW_ENV%\conda-meta" (
        ECHO No environment named "%CONDA_NEW_ENV%" exists, or is not a valid conda installation directory.
        EXIT /b 1
    )
    @REM already activated, don't change prompt
    @if "%CONDA_PREFIX%" == "%CONDA_NEW_ENV%" (
        @SET "CHANGE_PROMPT=0"
    ) else (
        @SET "CHANGE_PROMPT=1"
    )
    @SET "CONDA_PREFIX=%CONDA_NEW_ENV%"
    @REM manually construct the PATH, faster than ..activate
    @SET "NEW_PATH=%CONDA_NEW_ENV%;%CONDA_NEW_ENV%\Library\mingw-w64\bin;%CONDA_NEW_ENV%\Library\usr\bin;%CONDA_NEW_ENV%\Library\bin;%CONDA_NEW_ENV%\Scripts;%CONDA_SCRIPTS%"
    FOR /F "tokens=* delims=\" %%i IN ("%CONDA_NEW_ENV%") DO @SET "CONDA_NEW_ENV=%%~ni"
    @SET "CONDA_PS1_BACKUP=%PROMPT%"
    @goto skipchecks
)

@REM Ensure that path or name passed is valid before deactivating anything
@CALL "%CONDA_EXE%" ..checkenv "cmd.exe" "%CONDA_NEW_ENV%"
@IF errorlevel 1 exit /b 1

@REM The argument here tells the deactivate script to leave a placeholder for us when it removes PATH entries,
@REM    so that we can put our new path entries back in the same place
@call "%~dp0\deactivate.bat" "hold"
@if errorlevel 1 exit /b 1

@REM Activate the new environment
@FOR /F "delims=" %%i IN ('@call "%CONDA_EXE%" ..activate "cmd.exe" "%CONDA_NEW_ENV%"') DO @SET "NEW_PATH=%%i"
@IF errorlevel 1 exit /b 1

@REM take a snapshot of pristine state for later
@SET "CONDA_PS1_BACKUP=%PROMPT%"
@FOR /F "delims=" %%i IN ('@call "%CONDA_EXE%" ..changeps1') DO @SET "CHANGE_PROMPT=%%i"
@IF errorlevel 1 exit /b 1

:skipchecks
:: if our prompt var does not contain reference to CONDA_DEFAULT_ENV, set prompt
@IF "%CHANGE_PROMPT%" == "1" @IF "x%PROMPT:CONDA_DEFAULT_ENV=%" == "x%PROMPT%" (
    SET "PROMPT=(%CONDA_NEW_ENV%) %PROMPT%"
)

@REM always store the full path to the environment, since CONDA_DEFAULT_ENV varies
@FOR /F "tokens=1 delims=;" %%i in ("%NEW_PATH%") DO @SET "CONDA_PREFIX=%%i"

@REM Do we have CONDA_PATH_PLACEHOLDER in PATH?
@echo %PATH%|>nul findstr /rx ".*CONDA_PATH_PLACEHOLDER.*" && @set HAS_PLACEHOLDER=True || @set HAS_PLACEHOLDER=False

@REM look if the deactivate script left a placeholder for us.
@IF "%HAS_PLACEHOLDER%" == "True" (
    @REM If it did, replace it with our NEW_PATH
    @REM    Delayed expansion used here to do replacement with value of NEW_PATH
    @CALL SET "PATH=%%PATH:CONDA_PATH_PLACEHOLDER=!NEW_PATH!%%"
) ELSE (
    @REM If it did not, prepend NEW_PATH
    @SET "PATH=%NEW_PATH%;%PATH%"
)

@REM This persists env variables, which are otherwise local to this script right now.
@endlocal & (
    @REM Used for deactivate, to make sure we restore original state after deactivation
    @SET "CONDA_PS1_BACKUP=%CONDA_PS1_BACKUP%"
    @SET "PROMPT=%PROMPT%"
    @SET "PATH=%PATH%"
    @SET "CONDA_DEFAULT_ENV=%CONDA_PREFIX%"
    @SET "CONDA_PREFIX=%CONDA_PREFIX%"
    @SET "CONDA_SKIPCHECK="

    @REM Run any activate scripts
    @IF EXIST "%CONDA_PREFIX%\etc\conda\activate.d" (
        @PUSHD "%CONDA_PREFIX%\etc\conda\activate.d"
        @FOR %%g in (*.bat) DO @CALL "%%g"
        @POPD
    )
)

@goto :eof

:normalizepath
    @SET "%1=%~dpfn2"
    @EXIT /B
```
 
I hope this information is somewhat useful. Please let me know if I can help provide any additional details!

**Comment by t-kalinowski** [comment]
This link lists the recommended entry points for how to activate the environment: https://pro.arcgis.com/en/pro-app/latest/arcpy/get-started/installing-python-for-arcgis-pro.htm

After the break I'll try to install a trial version of ArcGIS to figure out the best way to integrate this with reticulate.

**Comment by philiporlando** [comment]
Thank you for linking these docs. I will review them and try to glean what I can. 

You would be going above and beyond if you were able to install a [trial version of ArcGIS Pro](https://www.esri.com/en-us/arcgis/products/arcgis-pro/trial). I really appreciate your support with this issue. Please let me know if there is anything else I can do to help troubleshoot in the meantime.

**Comment by philiporlando** [comment]
> Do you know how ESRI intends for this environment to be activated? Does running this command succeed in a cmd.exe terminal?
> 
> ```
> CALL "C:\Program Files\ArcGIS\Pro\bin\Python\envs\arcgispro-py3\Scripts\conda.bat" activate 
> ```

Realizing that my previous reply did not answer your question. Running this command returns the same `CommandNotFoundError: 'activate'` error.

**Comment by philiporlando** [comment]
This issue was resolved with PR #1195 and can be closed once this PR has been merged.
